---
title: "Check Haplotype LMM results"
output: NULL
author: Jinliang Yang
date: 08-27-2021
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
#library(tidyverse)
```

## LMM haplotype

```{bash}
module load bcftools
bcftools query -f '%CHROM\t%POS\t%REF[\t%SAMPLE=%DP]\n' Glu.hmp321_282_agpv4_maf005_miss03.vcf  
```




```{r}
# install.packages("haplo.stats")
library(haplo.stats)

 
library("data.table")
snp <- fread("data/09_282_block_hap/Glu.hmp321_282_agpv4_maf005_miss03.vcf", data.table = FALSE, skip="##")
names(snp)[1] <- "CHR"
snp$ID <- paste(snp$CHR, snp$POS, sep="-")

out <- tstrsplit(snp[1,10:ncol(snp)], "/", fixed=TRUE)
res <- data.frame(a1=out[[1]], a2=out[[2]])
id1 <- paste0(snp$ID[1], "_1")
id2 <- paste0(snp$ID[1], "_2")
names(res) <- c(id1, id2)
for(i in 2:nrow(snp)){
  out <- tstrsplit(snp[i,10:ncol(snp)], "/", fixed=TRUE)
  df <- data.frame(a1=out[[1]], a2=out[[2]])
  id1 <- paste0(snp$ID[i], "_1")
  id2 <- paste0(snp$ID[i], "_2")
  names(df) <- c(id1, id2)
  res <- cbind(res, df)
}

write.table(res, "cache/test.txt", sep="\t", row.names = FALSE, quote=FALSE)
```



```{r}
library(haplo.stats)
# https://cran.r-project.org/web/packages/haplo.stats/vignettes/haplostats.pdf

geno <- fread("cache/test.txt", header=T, data.table=FALSE)

geno[geno == "."] <- "NA"
em <- haplo.em(geno=geno[, 1:10], miss.val=NA)
```

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("gpart")
```

```{r}
library(gpart)

idx <- seq(from=1, to= ncol(geno), by=2)
g = geno[, idx[1:5]]
g[is.na(g)] <- 0
info <- snp[1:5, c("CHR", "ID", "POS")]
BigLDres = BigLD(geno = g, SNPinfo = info)
```

