---
title: "Chr5 gene expression level"
output: NULL
author: Jinliang Yang
date: 11-23-2020
---


```{r setup, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
#library(tidyverse)
```

## Study the DE between pre-60s and post-60s lines

Find the interesting candiate on Chr5 Zm00001d014451
GRMZM2G057459

Chr5:47,394,843..48,461,304

```{r}
library("data.table")

chr5 <- fread("cache/goi_chr5.csv", header=TRUE, data.table=FALSE)


early <- read.table("data/04Selection_scan/early_lines.txt")
late <- read.table("data/04Selection_scan/late_lines.txt")

df1 <- merge(early, chr5, by.x="V1", by.y="RNA_TaxaName")
df1$year <- "early"
df2 <- merge(late, chr5, by.x="V1", by.y="RNA_TaxaName")
df2$year <- "late"
df <- rbind(df1, df2)
```


## Metabolites

```{r}
m <- read.table("data/metabolite_rep1_genotype_namecorrect.txt", header=TRUE)

d <- merge(chr5, m, by.x="RNA_TaxaName", by.y="genotype")


idx <- grep("^X\\.", names(d))
out <- data.frame()
for(i in idx){
  #d$newd <- rescale(d[, i], to=c(0,1))
  fit <- lm(GRMZM2G148807 ~ d[, i] + TissueWODate, data=d)
  pval <- summary(fit)$coefficients[2,4]
  tem <- data.frame(trait= names(d)[i], pval=pval, eff=summary(fit)$coefficients[2,1], std=summary(fit)$coefficients[2,2])
  out <- rbind(out, tem)
}

out <- out[order(out$pval, decreasing=F),]

out1 <- subset(out, eff < 0)
out2 <- subset(out, eff > 0)

write.table(out1, "cache/exp_metabolite_neg.csv", sep=",", quote=FALSE, row.names=FALSE)
write.table(out2, "cache/exp_metabolite_pos.csv", sep=",", quote=FALSE, row.names=FALSE)
```



# Plot the results

```{r}
library(ggplot2)

df$TissueWODate <- factor(df$TissueWODate, levels = c( "GRoot", "GShoot", "L3Base", "L3Mid", "L3Tip", "LMAD", "LMAN", "Kern"),
                    labels =c( "Germinating Root", "Germinating Shoot", "Leaf 3 base","Leaf 3 Middle", "Leaf 3 tip", "Adult leaf during the day", "Adult leaf at night","Kernels after pollination"))

# df2 <- subset(avg, gene %in% c("GRMZM2G140559", "GRMZM2G054225", "RMZM2G037048", "GRMZM2G020148", "GRMZM2G000231"))


p <- ggplot(df, aes(x=TissueWODate, y=GRMZM2G148807, fill=year)) +
    geom_violin(trim=FALSE, position=position_dodge(1)) +
    xlab("") +
    ylab("avg(FPM)") +
    ggtitle("") +
    theme_classic() +
    #labs(color = "") +
    #scale_y_continuous(limits = c(0, 1.2)) +
    #geom_vline(xintercept=0, linetype="dashed", color = "red") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=18, face="bold"),
          #axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.position = c(0.8, 0.95), 
          axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.title = element_text(size=18, face="bold"),
          legend.text = element_text(size=18))
p


pdf("graphs/GRMZM2G057459_DE_by_tissue.pdf", width = 8, height= 6)
p
dev.off()
```


```{r}
tissue <- as.character(unique(df$TissueWODate))

outdf <- data.frame()
for(i in 1:length(tissue)){
  sub <- subset(df, TissueWODate %in% tissue[i])
  out <- t.test(subset(sub, year %in% "early")$GRMZM2G057459, subset(sub, year %in% "late")$GRMZM2G057459)
  tem <- data.frame(tissue=tissue[i], pval=out$p.value)
  outdf <- rbind(outdf, tem)
  
}

```

